src/services/schedule.js

```diff
@@ -0,0 +1,198 @@
+                    intStart.hour(startHr).minute(0).second(0).millisecond(0);
+                    let intEnd = intStart.clone();
+                    intEnd.hour(endHr).minute(0).second(0).millisecond(0);
+
+                    freeInts.push([intStart.format(), intEnd.format()]);
+                }
+            }
+            // We can't skip any days!
+            if (busyInt[0].date() > lastEndTime.date() + 1) {
+                lastEndTime.add(1, "days");
+                lastEndTime.hour(startHr).minute(0).second(0).millisecond(0);
+                console.log('rewind a day!');
+            } else
+                lastEndTime = busyInt[1];
+        });
+
+        return freeInts;
+    }
+
+    // Arguments:
+    //      tasks (Task Object) -> An array of task objects, as defined by the task model. 
+    //          Sorted from earliest to latest due date.
+    //      freeTime ([[integer, integer]]) -> An array of two-element arrays, each 
+    //          representing an interval of free time. The times should be in seconds since
+    //          epoch.
+    // Returns:
+    //      An array of Task objects with scheduled start times. If a task cannot be scheduled,
+    //      its start time is set to null.
+    const scheduleTasks = (tasks, freeTime) => {
+        if (tasks.length == 0)
+            return tasks;
+
+        let currTask = tasks[0]; // assume ordered by due date (earliest to latest)
+        let tDuration = currTask.duration * SECONDS_PER_MINUTE;
+
+        for (let i = 0; i < freeTime.length; i++) {
+            let freeInt = freeTime[i];
+            let iDuration = freeInt[1] - freeInt[0];
+
+            if (tDuration <= iDuration) {
+                // Set task start time to beginning of interval
+                currTask.scheduled_time = freeInt[0];
+
+                // Check that start_time + duration <= due_date
+                if (!isValidTask(currTask)) {
+                    // We can't schedule the event
+                    currTask.scheduled_time = null;
+                    return [currTask].concat(scheduleTasks(tasks.slice(1), freeTime));
+                }
+
+                let newFreeTime;
+                if (tDuration < iDuration) {
+                    let newInt = [freeInt[0] + tDuration, freeInt[1]];
+                    newFreeTime = [newInt].concat(freeTime.slice(0, i).concat(freeTime.slice(i + 1)));
+                } else if (tDuration == iDuration)
+                    newFreeTime = freeTime.slice(0, i).concat(freeTime.slice(i + 1))
+
+                // Sort newFreeTime
+                newFreeTime.sort((intA, intB) => {
+                    if (intA[0] < intB[0])
+                        return -1;
+                    else if (intA[0] == intB[0])
+                        return 0;
+                    else
+                        return 1;
+                });
+
+                if (newFreeTime[0][0] == currTask.scheduled_time)
+                    newFreeTime.splice(i, 1);
```

Comment:
This is where the bug in the scheduling algorithm was. Previously, I was removing the first element of `freeTime`, when it should have been the currently examined element.